src/discovery/dns/Akka.Discovery.Dns/Internal/AsyncDnsClient.cs

```diff
@@ -0,0 +1,421 @@
+        if (answer != null)
+        {
+            Sender.Tell(answer);
+            return;
+        }
+
+        SendDnsQuestion(id, name, recordType);
+    }
+
+    private void HandleDropRequest(DropRequest dropRequest)
+    {
+        var id = dropRequest.Question.Id;
+        if (_inflightRequests.TryGetValue(id, out var inFlight))
+        {
+            var sentQuestions = inFlight.Message.Questions.Select(q => new { q.Name, q.Type }).ToList();
+                
+            string expectedName = dropRequest.Question.Name;
+            DnsProtocol.RecordType expectedType = dropRequest.Question.RecordType;
+            
+            if (sentQuestions.Any(q => q.Name == expectedName && q.Type == expectedType))
+            {
+                _log.Debug("Dropping request [{0}]", id);
+                _inflightRequests.Remove(id);
+                Sender.Tell(new Dropped(id));
+            }
+            else if (_log.IsInfoEnabled)
+            {
+                _log.Info("Requested to drop request for id [{0}] expecting [{1}/{2}] but found requests for [{3}]... ignoring drop request",
+                    id, 
+                    expectedName, 
+                    expectedType,
+                    string.Join(", ", sentQuestions.Select(q => $"{q.Name}/{q.Type}")));
+            }
+        }
+        else
+        {
+            Sender.Tell(new Dropped(id));
+        }
+    }
+
+    internal virtual DnsProtocol.Message CreateMessage(string name, short id, DnsProtocol.RecordType recordType)
+    {
+        var question = new DnsProtocol.Question(name, recordType, DnsProtocol.RecordClass.In);
+        return new DnsProtocol.Message(
+            id,
+            new DnsProtocol.MessageFlags(),
+            ImmutableList.Create(question));
+    }
+
+    private IEnumerable<(string Name, DnsProtocol.RecordType Type)> WithAndWithoutTrailingDots(DnsProtocol.Question question)
+    {
+        yield return (question.Name, question.Type);
+            
+        if (question.Name.EndsWith("."))
+            yield return (question.Name.Substring(0, question.Name.Length - 1), question.Type);
+        else
+            yield return (question.Name + ".", question.Type);
+    }
+
+    private IActorRef CreateTcpClient()
+    {
+        var backoffOptions = Backoff.OnFailure(
+            childProps: Props.Create(() => new TcpDnsClient(_tcpManager, nameserver, Self)),
+            childName: "tcpDnsClient",
+            minBackoff: TimeSpan.FromMilliseconds(10),
+            maxBackoff: TimeSpan.FromSeconds(20),
+            randomFactor: 0.1, 
+            maxNrOfRetries: Int32.MinValue);
+            
+        return Context.ActorOf(
+            BackoffSupervisor.Props(backoffOptions),
+            "tcpDnsClientSupervisor");
+    }   
+    //TODO: Maybe this should be more resilient, what if we have a lot of requests at the same time?  
+    internal static short NewQueryId() => (short)Random.Next(0, 65535);
```

Comment:
This is a very bad way of doing this, the chance for ID collision is too high. 
The scala code gets around this collision problem by comparing the random number with the currently in-flight questions.